src/main/java/com/example/GymProject/service/TraineeService.java

```diff
@@ -1,56 +1,164 @@
+    }
 
-    public void createTrainee(String firstName, String lastName, Boolean isActive,
-                              LocalDate dob, String address, String userId) {
-        Trainee trainee = new Trainee(firstName, lastName, isActive, dob, address, userId);
-        boolean exists = usernameExists(memoryStorage.getUsernames(), trainee);
-        trainee.setUsername(generateUsername(trainee.getFirstName(), trainee.getLastName(), exists));
-        traineeDao.create(trainee);
+    //pordzel
+    public void deleteTraineeByUsername(String username, String password) {
+        Assert.notNull(username, "Username cannot be null");
+        authenticate(username, password);
+        traineeDao.deleteTraineeByUsername(username);
     }
 
-    public Trainee selectTrainee(String key) {
-        return memoryStorage.getTraineeDao().select(key);
+    public boolean matchUsernameAndPassword(String username, String password) {
+        Assert.notNull(username, "Username cannot be null");
+        Assert.notNull(password, "Password cannot be null");
+        return userService.matchUsernameAndPassword(username, password);
     }
 
-    public void updateTrainee(String key, Trainee newTrainee) {
-        Trainee trainee = new Trainee();
-        if (!memoryStorage.getTraineeDao().containsKey(key)) {
-            throw new NoSuchElementException("Wrong key, update failed!");
+
+    public void changeTraineePassword(String username, String newPassword, String password) {
+        Assert.notNull(username, "Username cannot be null");
+        Assert.notNull(newPassword, "New password cannot be null");
+        authenticate(username, password);
+        UserDto userDTO = userService.getUserByUsername(username);
+        if (userDTO != null) {
+            if (!password.equals(newPassword)) {
+                userDTO.setPassword(newPassword);
+                userService.updateUser(userDTO);
+                logger.info("Password has successfully changed for trainee {}", username);
+            } else {
+                logger.warn("Cannot change password for user {} since new password is equal to old password", username);
+
+            }
+        } else {
+            logger.warn("Cannot change password for user {} since such user was not found", username);
         }
-        boolean contains = usernameExists(memoryStorage.getUsernames(), newTrainee);
+    }
 
-        trainee.setUsername(generateUsername(newTrainee.getFirstName(), newTrainee.getLastName(), contains));
-        trainee.setPassword(selectTrainee(key).getPassword());
-        trainee.setFirstName(newTrainee.getFirstName());
-        trainee.setLastName(newTrainee.getLastName());
-        trainee.setAddress(newTrainee.getAddress());
-        trainee.setDob(newTrainee.getDob());
-        trainee.setActive(newTrainee.getActive());
+    public void activateDeactivateTrainee(String username, boolean isActive, String password) {
```

Comment:
Please have 2 special methods to exclude ambiguity.
`public void activate(String username, String password)`
`public void deactivate(String username, String password)`
And use `setIsActive(true)/setIsActive(false)` respectively. I will explain this during our next call.